### ISSUE TITLE
aws_s3_bucket_object is destroyed but not deleted

### ISSUE THREAD

**Issue by christianhaller** [terraform-providers/terraform-provider-aws#7262]
### Terraform Version

Terraform v0.11.11

- provider.aws v1.56.0

### Affected Resource(s)

* aws_s3_bucket_object
* aws_s3_bucket

### Terraform Configuration Files

```hcl
provider "aws" {
  region  = "eu-central-1"
}

resource "aws_s3_bucket" "bucket" {
  bucket = "s3-object-example"
  versioning {
    enabled = true //important for this example
  }
}

resource "aws_s3_bucket_object" "a" {
  bucket = "${aws_s3_bucket.bucket.bucket}"
  content = "a"
  key = "a"
}

resource "aws_s3_bucket_object" "b" {
  bucket = "${aws_s3_bucket.bucket.bucket}"
  content = "b"
  key = "/b" // 🚨 key is starting with slash
}

```

### Debug Output

```
aws_s3_bucket_object.a: Destroying... (ID: a)
aws_s3_bucket_object.b: Destroying... (ID: /b)
aws_s3_bucket_object.b: Destruction complete after 0s
aws_s3_bucket_object.a: Destruction complete after 1s

Apply complete! Resources: 0 added, 0 changed, 2 destroyed.
```

### Expected Behavior

Both s3 objects are deleted

### Actual Behavior

file b  is still in the bucket.
```bash
aws s3 ls s3://s3-object-example/
2019-01-23 13:55:01          1 b
```

### Steps to Reproduce

1. `terraform apply`
2. remove `resource "aws_s3_bucket_object "a"` and `resource "aws_s3_bucket_object "b"` from `main.tf`
3. `terraform apply` again

When the `key` is starting with a slash, the object is created, but not destroyed.

[![asciicast](https://asciinema.org/a/JgebKyvQyAVIr0wxtz9cxgFgT.svg)](https://asciinema.org/a/JgebKyvQyAVIr0wxtz9cxgFgT)

**Comment by ewbankkit** [comment]
The problem is because we are effectively ignoring any leading `/` in the S3 object's key - [`aws.Config`](https://docs.aws.amazon.com/sdk-for-go/api/aws/#Config)'s `DisableRestProtocolURICleaning` property is not set - and when listing object versions with that key we get none returned (and hence not deleted).
For backwards compatibility we cannot just set the `DisableRestProtocolURICleaning` property to `true` - `terraform plan` will start showing changes for `aws_s3_bucket_object` resources where `key` starts with '/', so we must handle this case explicitly in the object deletion code.

**Comment by christianhaller** [comment]
Thank you for your investigation @ewbankkit :)

**Comment by bflad** [comment]
Thanks to @ewbankkit the fix for this has been merged and will release with version 1.57.0 of the Terraform AWS provider, likely later today. 👍

**Comment by bflad** [comment]
This has been released in [version 1.57.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1570-january-26-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
